File: src/main/java/dev/jason/project/spring/vc_server/data/dto/DeviceDto.java
package dev.jason.project.spring.vc_server.data.dto;

import dev.jason.project.spring.vc_server.domain.model.Device;

import java.time.LocalDateTime;

public record DeviceDto(String name, Device.Type type, Device.OS os, String version, String token) {

    public Device toDomainModel(LocalDateTime lastTimeUsed) {
        return new Device(name, type, os, version, token, lastTimeUsed);
    }
    
    public static DeviceDto fromDomain(Device device) {
    	return new DeviceDto(device.name(), device.type(), device.os(), device.version(), device.token());
    }
}

File: core/src/main/java/dev/jason/app/compose/vaultchat/core/model/DeviceDto.kt
package dev.jason.app.compose.vaultchat.core.model

import kotlinx.serialization.Serializable

@Serializable
data class DeviceDto(
    val ownerUid: String,
    val name: String,
    val type: Type,
    val os: OS,
    val version: String,
    val token: String,
) {
    @Serializable
    enum class OS {
        Android;

        companion object {
            fun fromDomain(os: Device.Os) = when (os) {
                Device.Os.Android -> Android
            }
        }
    }

    @Serializable
    enum class Type {
        Mobile, Tablet;

        companion object {
            fun fromDomain(type: Device.Type) = when (type) {
                Device.Type.Mobile -> Mobile
                Device.Type.Tablet -> Tablet
            }
        }
    }
}

fun Device.toDto() = DeviceDto(
    ownerUid = ownerUid,
    name = name,
    type = DeviceDto.Type.fromDomain(type),
    os = DeviceDto.OS.fromDomain(os),
    version = version,
    token = token
)